libavcodec/x86/vvc/vvc_alf.asm

```diff
@@ -0,0 +1,801 @@
+    ; p6 = (y + i <= vb_pos + 2) ? p4 : p6;
+    cmp          vb_posq, %1
+    cmove            s1q, srcq
+    cmove            s2q, srcq
+
+    inc          vb_posq
+    cmp          vb_posq, %1
+    cmovae           s3q, s1q
+    cmovae           s4q, s2q
+
+    inc          vb_posq
+%if LUMA
+    cmp          vb_posq, %1
+    cmovae           s5q, s3q
+    cmovae           s6q, s4q
+%endif
+    sub          vb_posq, 2
+%%vb_end:
+%endmacro
+
+; SHIFT_VB(line)
+; brief: shift filter result
+; input:  m0, m1, vb_pos
+; output: m0
+; temp:   m9
+%macro SHIFT_VB 1
+%define SHIFT 7
+%if LUMA
+    cmp               %1, 3
+    je      %%near_above
+    cmp               %1, 0
+    je      %%near_below
+    jmp          %%no_vb
+    %%near_above:
+        cmp      vb_posq, 4
+        je     %%near_vb
+        jmp      %%no_vb
+    %%near_below:
+        cmp      vb_posq, 0
+        je     %%near_vb
+%else
+    cmp               %1, 0
+    je           %%no_vb
+    cmp               %1, 3
+    je           %%no_vb
+    cmp          vb_posq, 2
+    je         %%near_vb
+%endif
+%%no_vb:
+    vpsrad            m0, SHIFT
+    vpsrad            m1, SHIFT
+    jmp      %%shift_end
+%%near_vb:
+    vpbroadcastd      m9, [dd448]
+    paddd             m0, m9
+    paddd             m1, m9
+    vpsrad            m0, SHIFT + 3
+    vpsrad            m1, SHIFT + 3
+%%shift_end:
+    vpackssdw         m0, m0, m1
+%endmacro
+
+; FILTER_VB(line)
+; brief: filter pixels for luma and chroma
+; input:  line
+; output: m0, m1
+; temp:   s0q...s1q
+%macro FILTER_VB 1
+    vpbroadcastd      m0, [dw64]
+    vpbroadcastd      m1, [dw64]
+
+    GET_SRCS %1
+    %if LUMA
+        FILTER         0,  s5q,  s6q,  0 * ps
+        FILTER         1,  s3q,  s4q,  1 * ps
+        FILTER         2,  s3q,  s4q,  0 * ps
+        FILTER         3,  s3q,  s4q, -1 * ps
+        FILTER         4,  s1q,  s2q,  2 * ps
+        FILTER         5,  s1q,  s2q,  1 * ps
+        FILTER         6,  s1q,  s2q,  0 * ps
+        FILTER         7,  s1q,  s2q, -1 * ps
+        FILTER         8,  s1q,  s2q, -2 * ps
+        FILTER         9, srcq, srcq,  3 * ps
+        FILTER        10, srcq, srcq,  2 * ps
+        FILTER        11, srcq, srcq,  1 * ps
+    %else
+        FILTER         0,  s3q,  s4q,  0 * ps
+        FILTER         1,  s1q,  s2q,  1 * ps
+        FILTER         2,  s1q,  s2q,  0 * ps
+        FILTER         3,  s1q,  s2q, -1 * ps
+        FILTER         4, srcq, srcq,  2 * ps
+        FILTER         5, srcq, srcq,  1 * ps
+    %endif
+    SHIFT_VB %1
+%endmacro
+
+; LOAD_PIXELS(dest, src)
+%macro LOAD_PIXELS 2
+%if ps == 2
+    movu      %1, %2
+%else
+    vpmovzxbw %1, %2
+%endif
+%endmacro
+
+; STORE_PIXELS(dst, src)
+%macro STORE_PIXELS 2
+    %if ps == 2
+        movu         %1, m%2
+    %else
+        vpackuswb   m%2, m%2
+        vpermq      m%2, m%2, 0x8
+        movu         %1, xm%2
+    %endif
+%endmacro
+
+%macro FILTER_16x4 0
+    push clipq
```

Comment:
is it possible to get rid of all push and pop?

Reply:
I'm not sure but I'm getting a another try.

Reply:
If we try to calculate the next two pointers after the previous filter, we still need 4 pointers to use 2 for the filter and 2 to save the last pointer for the next cmp condition. And, we need another register for neg stride. Hence, we can only reduce one register from s1q,s2q,s3q,s4q,s5q,s6q to s1q, s2q, s3q, s4q,neg_stride, though, the codes get more duplicated. What do you think? 
``` ASM
%macro FILTER_VB 1
    vpbroadcastd      m0, [dw64]
    vpbroadcastd      m1, [dw64]

%define neg_strideq s6q
    mov  neg_strideq, src_strideq
    neg  neg_strideq

%if LUMA
    FILTER         9, srcq, srcq,  3 * ps
    FILTER        10, srcq, srcq,  2 * ps
    FILTER        11, srcq, srcq,  1 * ps

    cmp          vb_posq, 0
    je       %%vb_bottom
    cmp          vb_posq, 4
    jne         %%vb_end

%%vb_above:
    ; above
    ; p1 = (y + i == vb_pos - 1) ? p0 : p1;
    ; p2 = (y + i == vb_pos - 1) ? p0 : p2;
    ; p3 = (y + i >= vb_pos - 2) ? p1 : p3;
    ; p4 = (y + i >= vb_pos - 2) ? p2 : p4;
    ; p5 = (y + i >= vb_pos - 3) ? p3 : p5;
    ; p6 = (y + i >= vb_pos - 3) ? p4 : p6;
    lea          s3q, [srcq + src_strideq]
    lea          s4q, [srcq + neg_strideq]

    mov          s1q, s3q
    mov          s2q, s4q

    dec      vb_posq
    cmp      vb_posq, %1
    cmove        s1q, srcq
    cmove        s2q, srcq

    FILTER         4,  s1q,  s2q,  2 * ps
    FILTER         5,  s1q,  s2q,  1 * ps
    FILTER         6,  s1q,  s2q,  0 * ps
    FILTER         7,  s1q,  s2q, -1 * ps
    FILTER         8,  s1q,  s2q, -2 * ps

    lea          s3q, [s3q  + src_strideq]
    lea          s4q, [s4q  + neg_strideq]

    dec      vb_posq
    cmp      vb_posq, %1
    cmova        s1q, s3q
    cmova        s2q, s4q

    FILTER         1,  s1q,  s2q,  1 * ps ; FILTER         1,  s3q,  s4q,  1 * ps
    FILTER         2,  s1q,  s2q,  0 * ps ; FILTER         2,  s3q,  s4q,  0 * ps
    FILTER         3,  s1q,  s2q, -1 * ps ; FILTER         3,  s3q,  s4q, -1 * ps

    lea          s3q, [s3q  + src_strideq]
    lea          s4q, [s4q  + neg_strideq]
    dec      vb_posq
    cmp      vb_posq, %1
    cmova        s1q, s3q
    cmova        s2q, s4q
    FILTER         0, s1q,  s2q,  0 * ps

    add      vb_posq, 3
    jmp     %%shift_vb

%%vb_bottom:
    ; bottom
    ; p1 = (y + i == vb_pos    ) ? p0 : p1;
    ; p2 = (y + i == vb_pos    ) ? p0 : p2;
    ; p3 = (y + i <= vb_pos + 1) ? p1 : p3;
    ; p4 = (y + i <= vb_pos + 1) ? p2 : p4;
    ; p5 = (y + i <= vb_pos + 2) ? p3 : p5;
    ; p6 = (y + i <= vb_pos + 2) ? p4 : p6;
    lea          s3q, [srcq + src_strideq]
    lea          s4q, [srcq + neg_strideq]

    mov          s1q, s3q
    mov          s2q, s4q

    cmp      vb_posq, %1
    cmove        s1q, srcq
    cmove        s2q, srcq

    FILTER         4,  s1q,  s2q,  2 * ps
    FILTER         5,  s1q,  s2q,  1 * ps
    FILTER         6,  s1q,  s2q,  0 * ps
    FILTER         7,  s1q,  s2q, -1 * ps
    FILTER         8,  s1q,  s2q, -2 * ps

    lea          s3q, [s3q  + src_strideq]
    lea          s4q, [s4q  + neg_strideq]

    inc      vb_posq
    cmp      vb_posq, %1
    cmovb        s1q, s3q
    cmovb        s2q, s4q

    FILTER         1,  s1q,  s2q,  1 * ps ; FILTER         1,  s3q,  s4q,  1 * ps
    FILTER         2,  s1q,  s2q,  0 * ps ; FILTER         2,  s3q,  s4q,  0 * ps
    FILTER         3,  s1q,  s2q, -1 * ps ; FILTER         3,  s3q,  s4q, -1 * ps

    lea          s3q, [s3q  + src_strideq]
    lea          s4q, [s4q  + neg_strideq]
    inc      vb_posq
    cmp      vb_posq, %1
    cmovb        s1q, s3q
    cmovb        s2q, s4q
    FILTER         0, s1q,  s2q,  0 * ps

    sub      vb_posq, 2
    jmp   %%shift_vb
%%vb_end:
    lea          s1q, [srcq  + src_strideq]
    lea          s2q, [srcq  + neg_strideq]

    FILTER         4,  s1q,  s2q,  2 * ps
    FILTER         5,  s1q,  s2q,  1 * ps
    FILTER         6,  s1q,  s2q,  0 * ps
    FILTER         7,  s1q,  s2q, -1 * ps
    FILTER         8,  s1q,  s2q, -2 * ps

    lea          s1q, [s1q  + src_strideq]
    lea          s2q, [s2q  + neg_strideq]
    FILTER         1,  s1q,  s2q,  1 * ps
    FILTER         2,  s1q,  s2q,  0 * ps
    FILTER         3,  s1q,  s2q, -1 * ps

    lea          s1q, [s1q  + src_strideq]
    lea          s2q, [s2q  + neg_strideq]
    FILTER         0, s1q,  s2q,  0 * ps

%else
    GET_SRCS %1
    FILTER         0,  s3q,  s4q,  0 * ps
    FILTER         1,  s1q,  s2q,  1 * ps
    FILTER         2,  s1q,  s2q,  0 * ps
    FILTER         3,  s1q,  s2q, -1 * ps
    FILTER         4, srcq, srcq,  2 * ps
    FILTER         5, srcq, srcq,  1 * ps
%endif
%%shift_vb:
    SHIFT_VB %1
%endmacro
```

Reply:
Just add some macro and the chroma will use only one push/pop
``` ASM
%macro FILTER_16x4 0
%if LUMA
    push clipq
    push strideq
    %define s5q clipq
    %define s6q strideq
%endif
    %define s3q pixel_maxq
    %define s4q offsetq

    push xq

    xor               xq, xq
%%filter_16x4_loop:
    LOAD_PIXELS       m2, [srcq]   ;p0

    FILTER_VB         xq

    paddw             m0, m2

    ;clip to pixel
    CLIPW             m0, m14, m15

    STORE_PIXELS  [dstq], 0

    lea             srcq, [srcq + src_strideq]
    lea             dstq, [dstq + dst_strideq]
    inc               xq
    cmp               xq, 4
    jl %%filter_16x4_loop

    mov               xq, src_strideq
    neg               xq
    lea             srcq, [srcq + xq * 4]
    mov               xq, dst_strideq
    neg               xq
    lea             dstq, [dstq + xq * 4]

    pop xq
%if LUMA
    pop strideq
    pop clipq
%endif
%endmacro
```

Reply:
Using `push xq` will be faster and have less memory usage than REP 4
```
before: 34.5 KB (35,361 Bytes)
ff_vvc_alf_filter_luma_12_avx2 1368
ff_vvc_alf_filter_luma_8_avx2 1566

after: 15.5 KB (15,889 Bytes)
ff_vvc_alf_filter_luma_12_avx2 1343
ff_vvc_alf_filter_luma_8_avx2 1563
```

Reply:
Okay! Test all cases and send the most optimized patches.